### ISSUE TITLE
identifier_uri must start with "api://, https://" 

### ISSUE THREAD

**Issue by philipp1992** [hashicorp/terraform-provider-azuread#951]
When creating AppRegistrations, the identifier_uri must start with "api://, https://" and so on, unfortunately not all applications support this prefix at the service provider side. Can you please add a method to skip this validation, so we can use arbitrary identifier uris.

kind regards
Philipp

```
	"identifier_uris": {
				Type:     schema.TypeList,
				Optional: true,
				Computed: true,
				Elem: &schema.Schema{
					Type:         schema.TypeString,
					ValidateFunc: validate.URLIsAppURI,
				},
			},

```

**Comment by manicminer** [comment]
Hi @philipp1992, thanks for reporting this. Do you have an example of an identifier URI you'd like to set that is failing our validation? We can definitely look to expanding the validation however it may be that the API does not accept the value, in which case changing the validation will not help you. Thank!

**Comment by manicminer** [comment]
Unfortunately a domain-only identifier URI isn't actually supported by Microsoft Graph - see the following screenshot:

<img width="763" alt="Screenshot 2023-11-27 at 16 24 40" src="https://github.com/hashicorp/terraform-provider-azuread/assets/251987/6e9cff98-300e-425b-8f7b-dd5ba9d2598e">

If this is supported in the portal, then this is unfortunately not extended to the public API which means we cannot support it. We do aim to match the validation of the API, so if this changes then we'll update the provider validation accordingly.

Since there is nothing that we can do here, I'm going to close this issue for now. However, if the API gets updated, please feel free to comment further, or open a new issue, and we'll be happy to look at it again.

**Comment by alex-nutu** [comment]
Hello, thanks for the quick response, but as mentioned, it is possible via MsGraph, using the PATCH method

Documentation https://learn.microsoft.com/en-us/graph/api/application-update?view=graph-rest-1.0&tabs=http

Before
![image](https://github.com/hashicorp/terraform-provider-azuread/assets/127733267/11a3e6ab-6769-499c-bf64-63c27c39bb13)

After the patch
![image](https://github.com/hashicorp/terraform-provider-azuread/assets/127733267/0aa066d0-cae1-49ad-a87b-0afbe927ad68)
![image](https://github.com/hashicorp/terraform-provider-azuread/assets/127733267/70fa5519-0fdf-4d5e-8652-776a8e05bcdd)
![image](https://github.com/hashicorp/terraform-provider-azuread/assets/127733267/44fd2bb6-de5f-4d46-8401-32af84568943)

![image](https://github.com/hashicorp/terraform-provider-azuread/assets/127733267/0b26701e-71fa-4d59-980f-91ac5396e323)

Can you please re-open the issue? 
Thank you

**Comment by Tbohunek** [comment]
@manicminer can you reopen the issue or we need to wait for @philipp1992 ?

Either way, I hit the same issue, and I also suspect it might not be possible to fix at all - is TF able to differentiate between PUT and PATCH to append such configuration?

Whether or not should have MS allowed this with PATCH is another discussion, but we have 3rd Party App integrations that we simply can't provision with Terraform, and that kinda put the spanner in the wheels of the "move to TFE" bandwagon.

**Comment by manicminer** [comment]
Thanks for reporting back. I've reopened this although I suspect there's more going on here because it still doesn't work for me.

<img width="809" alt="Screenshot 2023-11-27 at 18 22 46" src="https://github.com/hashicorp/terraform-provider-azuread/assets/251987/d7c0dd8e-3649-471b-bdb4-cf836394ae40">

Is `abc.eu` a verified domain of your tenant? Do you need to first set `https://abc.eu` before patching it to be `abc.eu`? I've tried this and it doesn't seem to work either way for me. Any other context you can provide will be helpful!

**Comment by alex-nutu** [comment]
Hi, `abc.eu` is not a verified domain. Indeed, TF is configured to use `https://abc.eu` as initial deployment. If not, the URI error is present.

After the application is created, i use GraphExplorer to change the `identifier`

The process is:
1. apply without `identifierUris` attribute, because of the following error ` ApplicationsClient.BaseClient.Post(): unexpected status 400 with OData error: HostNameNotOnVerifiedDomain: Values of identifierUris property must use a verified domain of the organization or its subdomain: 'https://abc.eu'`
2. add the attribute `identifierUris` in tf code in format `https://abc.eu` and run another apply
3. If i try another apply with `identifierUris` in format `abc.eu` i get the `Error: URI has no host`
4. Use MsGraph with the patch method as explained before

Hopefully this will provide the full picture, if not let me know what other info i can provide.

I appreciate your help.

**Comment by manicminer** [comment]
@alex-nutu Can you post your full application resource config? I am unable to patch an application with an unverified domain (i.e. step 2 in your repro guide):

<img width="343" alt="Screenshot 2023-11-28 at 11 33 10" src="https://github.com/hashicorp/terraform-provider-azuread/assets/251987/bbd2f5e3-5522-4594-8e3a-16c06a7c09e9">
  
  
<img width="697" alt="Screenshot 2023-11-28 at 11 32 18" src="https://github.com/hashicorp/terraform-provider-azuread/assets/251987/62aa95c6-9e45-413f-b5f4-405233d702af">

**Comment by alex-nutu** [comment]
The deployment is using a module where the service principal is also created. I've tried to replicate the issue without the module and i encounter the same issue. So will need to check inside the module what could cause this.

However, this is not the topic of the issue, because i've used one of the domain we have added to AzureAD under `custom domain names` and i have the same error `Error: URI has no host for "identifier_uris.0"`

Steps to replicate 

```
resource "azuread_application" "test" {
  display_name    = "issue951"
  identifier_uris = ["https://abc.<custom.domain>"]
}
```

Apply and then replace the code with 


```
resource "azuread_application" "test" {
  display_name    = "issue951"
  identifier_uris = ["abc.<custom.domain>"]
}
```
